src/mono/mono.proj

```diff
@@ -111,17 +114,74 @@
           DestinationFolder="$([System.IO.Directory]::GetDirectories('$(LocalMonoDotnetDir)\shared\Microsoft.NETCore.App')[0])" />
   </Target>
 
+
+  <!-- Precompile (AOT) managed libs in .dotnet-mono\shared\Microsoft.NETCore.App\**\*.dll -->
+  <Target Name="PrecompileLocalMonoDotnetBcl" DependsOnTargets="PrepareLocalMonoDotnetBclForAot">
+    <CallTarget Targets="PrecompileAssembly"/>
+  </Target>
+  <Target Name="PrepareLocalMonoDotnetBclForAot" DependsOnTargets="PatchLocalMonoDotnet">
+    <PropertyGroup>
+      <MonoExec>$(LocalMonoDotnet)</MonoExec>
+    </PropertyGroup>
+    <ItemGroup>
+      <AssembliesToPrecompile 
+          Include="$(LocalMonoDotnetDir)\shared\Microsoft.NETCore.App\**\*.dll"
+          Exclude="$(LocalMonoDotnetDir)\shared\Microsoft.NETCore.App\**\System.Runtime.WindowsRuntime.dll" />
+    </ItemGroup>
+  </Target>
+
+  <!-- Precompile (AOT) $(ArtifactsDir)bin\**\*Tests\**\System.*.dll -->
+  <Target Name="PrecompileCoreFxTests" DependsOnTargets="PrepareCoreFxTestsForAot">
+    <CallTarget Targets="PrecompileAssembly"/>
+  </Target>
+  <Target Name="PrepareCoreFxTestsForAot" DependsOnTargets="PatchCoreFxTesthost">
+    <PropertyGroup>
+      <MonoExec>$(CoreFxTesthost)dotnet</MonoExec>
+    </PropertyGroup>
+    <ItemGroup>
+      <AssembliesToPrecompile
+          Include="$(ArtifactsDir)bin\**\*Tests\**\System.*.dll"
+          Exclude="$(ArtifactsDir)bin\**\System.Utf8String.Experimental.Tests.dll" />
+    </ItemGroup>
+  </Target>
+  
+  <!-- Precompile (AssembliesToPrecompile) using (MonoExec) -->
+  <Target Name="PrecompileAssembly">
+   <PropertyGroup>
+      <EnvVars Condition="'$(OS)' != 'Windows_NT'">
+      <![CDATA[MONO_ENV_OPTIONS="--aot=mcpu=native"]]>
+      </EnvVars>
+      <EnvVars Condition="'$(OS)' == 'Windows_NT'">
+      <![CDATA[set "MONO_ENV_OPTIONS=--aot=mcpu=native"]]>
+      </EnvVars>
+    </PropertyGroup>
+    <Exec Command="$(EnvVars) $(MonoExec) %(AssembliesToPrecompile.Identity)" ContinueOnError="WarnAndContinue" />
```

Comment:
I'd be really great if we could avoid doing it this way and use proper publish step

Reply:
I agree but it's at the moment it's the only way, but already looking for better approaches.